src/compute/src/render/top_k.rs

```diff
@@ -235,62 +235,62 @@ where
+                            target.push(((*row).clone(), diff.clone()));
+                        }
+                        // local copies that may count down to zero.
+                        let mut offset = offset;
+                        let mut limit = limit;
+
+                        // The order in which we should produce rows.
+                        let mut indexes = (0..source.len()).collect::<Vec<_>>();
+                        // We decode the datums once, into a common buffer for efficiency.
+                        // Each row should contain `arity` columns; we should check that.
+                        let mut buffer = Vec::with_capacity(arity * source.len());
+                        for (index, row) in source.iter().enumerate() {
+                            buffer.extend(row.0.iter());
+                            assert_eq!(buffer.len(), arity * (index + 1));
+                        }
+                        let width = buffer.len() / source.len();
+
+                        //todo: use arrangements or otherwise make the sort more performant?
+                        indexes.sort_by(|left, right| {
+                            let left = &buffer[left * width..][..width];
+                            let right = &buffer[right * width..][..width];
+                            // Note: source was originally ordered by the u8 array representation
+                            // of rows, but left.cmp(right) uses Datum::cmp.
+                            mz_expr::compare_columns(&order_key, left, right, || left.cmp(right))
+                        });
+
+                        // We now need to lay out the data in order of `buffer`, but respecting
+                        // the `offset` and `limit` constraints.
+                        for index in indexes.into_iter() {
+                            let (row, mut diff) = source[index];
+                            if diff > 0 {
+                                // If we are still skipping early records ...
+                                if offset > 0 {
+                                    let to_skip =
+                                        std::cmp::min(offset, usize::try_from(diff).unwrap());
+                                    offset -= to_skip;
+                                    diff -= Diff::try_from(to_skip).unwrap();
+                                }
+                                // We should produce at most `limit` records.
+                                // TODO(benesch): avoid dangerous `as` conversion.
+                                #[allow(clippy::as_conversions)]
+                                if let Some(limit) = &mut limit {
+                                    diff = std::cmp::min(diff, Diff::try_from(*limit).unwrap());
+                                    *limit -= diff as usize;
+                                }
+                                // Output the indicated number of rows.
                                 if diff > 0 {
```

Comment:
You could be more aggressive with avoiding rightward-drift by `continue`ing on the diff checks. I'd personally prefer this but I know different people think differently, so I'll leave that up to you and @vmarcos.

Reply:
Good call; this is my preference too, and it would be nice to renovate this code while I'm touching it.